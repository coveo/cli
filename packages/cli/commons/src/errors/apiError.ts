import {red} from 'chalk';
import dedent from 'ts-dedent';
import {PrintableError, SeverityLevel} from './printableError';

export interface APIErrorResponse {
  message: string;
  errorCode: string;
  requestID: string;
}

export interface AxiosErrorFromAPI {
  response: {
    status: number;
    data: APIErrorResponse;
  };
}

export const APIErrorSchema = {
  type: 'object',
  required: ['errorCode', 'message'],
  properties: {
    errorCode: {
      type: 'string',
    },
    message: {
      type: 'string',
    },
  },
};

export const AxiosErrorFromAPISchema = {
  type: 'object',
  required: ['response'],
  properties: {
    response: {
      type: 'object',
      required: ['status', 'data'],
      properties: {
        status: {
          type: 'integer',
        },
        data: APIErrorSchema,
      },
    },
  },
};

export class APIError extends PrintableError {
  public constructor(
    error: APIErrorResponse | AxiosErrorFromAPI,
    tagLine?: string
  ) {
    super(SeverityLevel.Error);
    let errorCode: string;
    let message: string;
    let status: number | undefined;
    if (this.isFromAxios(error)) {
      errorCode = error.response.data.errorCode;
      message = error.response.data.message;
      status = error.response.status;
    } else {
      errorCode = error.errorCode;
      message = error.message;
    }

    this.name = `APIError - ${errorCode}`;
    this.message = dedent`
    ${tagLine}
    ${status ? `Status code: ${status}\n` : ''}
    Error code: ${red(errorCode)}
    Message: ${red(message)}
    `;
  }

  private isFromAxios(
    error: APIErrorResponse | AxiosErrorFromAPI
  ): error is AxiosErrorFromAPI {
    return 'response' in error;
  }
}
