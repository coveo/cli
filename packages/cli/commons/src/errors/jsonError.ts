import type {ValidationError} from 'jsonschema';

export function getPrettyJsonValidationErrors(
  errors: ValidationError[]
): string {
  const stackToPrettyError = (stack: string) =>
    ` - ${stack.replace(/^instance\./, '')}\n`;
  return errors
    .reduce<string>(
      (errors, error) => (errors += stackToPrettyError(error.stack)),
      '\n'
    )
    .replace(/\n$/, '');
}
