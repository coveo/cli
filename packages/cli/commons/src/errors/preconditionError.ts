import {PrintableError, SeverityLevel} from './printableError';

export interface PreconditionErrorOptions {
  category?: string;
  level?: SeverityLevel;
}

export class PreconditionError extends PrintableError {
  public constructor(
    public message: string,
    public options?: PreconditionErrorOptions
  ) {
    super(options?.level || SeverityLevel.Error);
    this.name = 'Precondition Error';
    if (this.options?.category) {
      this.name += ` - ${this.options?.category}`;
    }
  }
}
