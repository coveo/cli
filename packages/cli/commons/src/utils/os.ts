export const appendCmdIfWindows = (cmd: TemplateStringsArray) =>
  `${cmd}${process.platform === 'win32' ? '.ps1' : ''}`;
