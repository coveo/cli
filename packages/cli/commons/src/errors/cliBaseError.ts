import {CLIError} from '@oclif/core/lib/errors';
import {Chalk, red, yellow, cyan} from 'chalk';

export enum SeverityLevel {
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
}

export interface CLIBaseErrorInterface {
  level?: SeverityLevel;
  cause?: Error;
}

interface OClifCLIError extends Omit<CLIError, 'render' | 'bang' | 'oclif'> {}

export class CLIBaseError extends Error implements OClifCLIError {
  private static readonly defaultSeverity: SeverityLevel = SeverityLevel.Error;
  public name = 'CLI Error';

  public constructor(
    private error?: string | Error | CLIError,
    private options?: CLIBaseErrorInterface
  ) {
    super(error instanceof Error ? error.message : error, options);
  }

  public get stack(): string {
    return super.stack || '';
  }

  public get severityLevel(): SeverityLevel {
    return this.options?.level || CLIBaseError.defaultSeverity;
  }

  /**
   * Specific to internal oclif error handling
   */
  private get oclif() {
    return this.error instanceof CLIError ? this.error.oclif : {};
  }

  /**
   * Used by oclif to pretty print the error
   */
  private get bang() {
    let color: Chalk;

    switch (this.severityLevel) {
      case SeverityLevel.Info:
        color = cyan;
        break;

      case SeverityLevel.Warn:
        color = yellow;
        break;

      case SeverityLevel.Error:
      default:
        color = red;
        break;
    }
    return color('Â»');
  }
}
