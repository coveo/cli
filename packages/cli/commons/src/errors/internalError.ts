import {CLIBaseError} from './cliBaseError';
import {dedent} from 'ts-dedent';

export class InternalError extends CLIBaseError {
  public name = 'Internal CLI Error';
  public constructor(message: string) {
    super();
    this.message = dedent`
      Internal CLI Error, please report to Coveo
      
      Message:
        ${message}`;
  }
}
