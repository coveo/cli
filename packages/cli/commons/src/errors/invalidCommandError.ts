import {CLIBaseError, SeverityLevel} from './cliBaseError';
import dedent from 'ts-dedent';

export class InvalidCommandError extends CLIBaseError {
  public name = 'Invalid CLI Command';
  public constructor(cause: string) {
    super('', {level: SeverityLevel.Error});
    this.message = dedent`
      An error has been found in the CLI, please report this error to Coveo.
      
      Cause:
      ${this.cause}

      Stack:
      ${this.stack}
      `;
  }
}
