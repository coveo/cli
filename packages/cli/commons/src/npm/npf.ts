import {fork, ForkOptions, ChildProcess, spawnSync} from 'node:child_process';
import globalConfig from '../config/globalConfig';
import {join, resolve} from 'node:path';
import {existsSync, mkdirSync} from 'node:fs';
import {createRequire} from 'node:module';
import npa from 'npm-package-arg';
import {InternalError} from '../errors/internalError';
import {appendCmdIfWindows} from '../utils/os';
import {satisfies} from 'semver';

/**
 * {@see https://nodejs.org/api/errors.html#module_not_found}
 */
const NODE_ERR_MODULE_NOT_FOUND = 'MODULE_NOT_FOUND';

interface Pjson {
  version: string;
  main?: string;
  exports?: string;
}

/**
 * Node Package Fork: Do NPX, but with a `child_process.fork`.
 * @param packageSpec the package specifier, {@see https://docs.npmjs.com/cli/v9/using-npm/package-spec?v=true}
 * @param opts {NpfOptions} tweak some options for fork & npf
 * @returns {ChildProcess} the forked process.
 */
export default function (
  packageSpec: string,
  args: string[] = [],
  opts?: ForkOptions
): ChildProcess {
  if (!isPackageInstalled(packageSpec)) {
    installPackage(packageSpec);
  }
  const entrypoint = getEntrypointFromPackageSpec(packageSpec);
  return fork(entrypoint, args, opts);
}

function isPackageInstalled(packageSpec: string): boolean {
  const {name: packageName, saveSpec: packageVersion} = npa(packageSpec);
  if (!packageName) {
    throw new InternalError(`Package spec ${packageSpec} seems invalid`);
  }
  let packageJsonFromRequire = getPackageJsonFromLazyLoadDirectory(packageName);
  return packageVersion
    ? satisfies(packageJsonFromRequire?.version ?? '', packageVersion)
    : Boolean(packageJsonFromRequire);
}

function installPackage(packageSpec: string) {
  if (!existsSync(join(getLazyLoadedDependencyProjectPath()))) {
    ensureCliNodeModuleFolderExists();
    initializeNpmProject();
  }
  spawnSync(appendCmdIfWindows`npm`, ['install', '-E', packageSpec], {
    cwd: getLazyLoadedDependencyDirectoryPath(),
    shell: process.platform === 'win32' ? 'powershell' : undefined,
  });
}

function ensureCliNodeModuleFolderExists() {
  mkdirSync(getLazyLoadedDependencyDirectoryPath(), {
    recursive: true,
  });
}

function getLazyLoadedDependencyProjectPath(): string {
  return join(getLazyLoadedDependencyDirectoryPath(), 'package.json');
}

function getLazyLoadedDependencyDirectoryPath(): string {
  return resolve(globalConfig.get().dataDir, 'lazyLoadedDependencies');
}

function getPackageJsonFromLazyLoadDirectory(
  packageName: string
): Pjson | null | never {
  let packageJson = null;
  const packageJsonPath = join(packageName, 'package.json');
  try {
    packageJson = requireFromLazy()(packageJsonPath);
  } catch (error) {
    if (isErrorWithCode(error) && error.code !== NODE_ERR_MODULE_NOT_FOUND) {
      throw error;
    }
  }
  return packageJson;
}

function initializeNpmProject() {
  spawnSync(appendCmdIfWindows`npm`, ['init', '-y'], {
    cwd: getLazyLoadedDependencyDirectoryPath(),
    shell: process.platform === 'win32' ? 'powershell' : undefined,
  });
}

function getEntrypointFromPackageSpec(packageSpec: string): string {
  const {name: packageName} = npa(packageSpec);
  if (!packageName) {
    throw new InternalError(`Package spec ${packageSpec} seems invalid`);
  }
  try {
    return requireFromLazy().resolve(packageName);
  } catch (error) {
    if (isErrorWithCode(error) && error.code === NODE_ERR_MODULE_NOT_FOUND) {
      throw new InternalError(
        `Package ${packageName} does not have an entrypoint.`
      );
    } else {
      throw error;
    }
  }
}

function requireFromLazy() {
  return createRequire(getLazyLoadedDependencyProjectPath());
}

function isErrorWithCode(error: any): error is Error & {code: string} {
  return typeof error['code'] === 'string' && error instanceof Error;
}
