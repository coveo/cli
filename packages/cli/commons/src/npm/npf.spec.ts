import type {Interfaces} from '@oclif/core';
jest.mock('node:fs');
import {existsSync, mkdirSync} from 'node:fs';
jest.mock('node:child_process');
import {fork, spawnSync} from 'node:child_process';
jest.mock('node:module');
import {createRequire} from 'node:module';
jest.mock('npm-package-arg');
import npa from 'npm-package-arg';
jest.mock('../config/globalConfig');
import globalConfig from '../config/globalConfig';
jest.mock('../utils/os');
import {appendCmdIfWindows} from '../utils/os';
import npf from './npf';
import {resolve} from 'node:path';
import {Result} from 'npm-package-arg';
import {InternalError} from '../errors/internalError';

describe('npf', () => {
  const mockedExistSync = jest.mocked(existsSync);
  const mockedMkdirSync = jest.mocked(mkdirSync);
  const mockedFork = jest.mocked(fork);
  const mockedSpawnSync = jest.mocked(spawnSync);
  const mockedNpa = jest.mocked(npa);
  const mockedAppendCmdIfWindows = jest.mocked(appendCmdIfWindows);
  const mockedGlobalConfig = jest.mocked(globalConfig);
  const mockedCreateRequire = jest.mocked(createRequire);
  let mockedRequire: jest.Mock;
  const mockedRequireResolve = jest.fn();
  const doThrowModuleNotFound = () => {
    const error = new Error();
    Object.assign(error, {code: 'MODULE_NOT_FOUND'});
    throw error;
  };
  const fakeLazyLoadedDepFolderPath = resolve(
    '/i/am/groot/',
    'lazyLoadedDependencies'
  );

  beforeEach(() => {
    mockedGlobalConfig.get.mockReturnValue({
      dataDir: '/i/am/groot/',
    } as Interfaces.Config);
    mockedAppendCmdIfWindows.mockImplementation((input) => `${input}`);
    mockedRequire = jest.fn();
    mockedCreateRequire.mockImplementation(() => {
      Object.assign(mockedRequire, {resolve: mockedRequireResolve});
      return mockedRequire as unknown as NodeRequire;
    });
    mockedRequireResolve.mockImplementation(() => '/some/silly/path.js');
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('when the packageName can be resolved by npa', () => {
    beforeEach(() => {
      mockedNpa.mockReturnValue({
        name: 'potato',
        saveSpec: '1.2.3',
      } as unknown as Result);
    });

    it('should create a require from the lazyLoaded package.json and resolve the name of the package', () => {
      npf('potato@1.2.3');
      expect(mockedRequireResolve).toBeCalledWith('potato');
    });

    describe('when the lazyLoad npm project does not exists', () => {
      beforeEach(() => {
        mockedExistSync.mockReturnValue(false);
        mockedCreateRequire.mockImplementationOnce(doThrowModuleNotFound);
      });

      it('should initialize the project and install the dependency', () => {
        npf('potato@1.2.3');

        expect(mockedMkdirSync).toBeCalledWith(fakeLazyLoadedDepFolderPath, {
          recursive: true,
        });
        expect(mockedSpawnSync).toHaveBeenNthCalledWith(
          1,
          'npm',
          ['init', '-y'],
          {cwd: fakeLazyLoadedDepFolderPath}
        );
        expect(mockedSpawnSync).toHaveBeenNthCalledWith(
          2,
          'npm',
          ['install', '-E', 'potato@1.2.3'],
          {cwd: fakeLazyLoadedDepFolderPath}
        );
      });
    });

    describe('when the lazyLoad npm project does exists', () => {
      beforeEach(() => {
        mockedExistSync.mockReturnValue(true);
      });

      describe('when the dependency is missing', () => {
        beforeEach(() => {
          mockedRequire.mockImplementationOnce(() => null);
        });

        it.each([
          {
            packageSpec: 'potato',
            packageName: 'potato',
            itName:
              'should initialize the project and install the dependency when there is no version requirement',
          },
          {
            packageSpec: 'potato@1.2.3',
            packageName: 'potato',
            packageVersion: '1.2.3',
            itName:
              'should initialize the project and install the dependency when there is a version requirement',
          },
        ])('$itName', ({packageSpec, packageName, packageVersion}) => {
          mockedNpa.mockReturnValueOnce({
            name: packageName,
            saveSpec: packageVersion,
          } as unknown as Result);

          npf(packageSpec);

          expect(mockedSpawnSync).toHaveBeenNthCalledWith(
            1,
            'npm',
            ['install', '-E', packageSpec],
            {cwd: fakeLazyLoadedDepFolderPath}
          );
        });
      });

      describe('when the dependency is installed', () => {
        beforeEach(() => {
          mockedRequire.mockImplementationOnce(() => ({
            version: '1.0.0',
          }));
        });

        describe('when there is no version requirement', () => {
          beforeEach(() => {
            mockedNpa.mockReturnValue({
              name: 'potato',
            } as unknown as Result);
          });

          it('should forked the resolved entrypoint without install', () => {
            npf('potato');
            expect(mockedSpawnSync).not.toBeCalled();
            expect(mockedFork).toHaveBeenNthCalledWith(
              1,
              '/some/silly/path.js',
              [],
              undefined
            );
          });
        });

        describe('when there is a version requirement', () => {
          describe('when the version requirement is satisfied', () => {
            beforeEach(() => {
              mockedNpa.mockReturnValue({
                name: 'potato',
                saveSpec: '1.0.0',
              } as unknown as Result);
            });

            it('should fork using the resolved entrypoint', () => {
              npf('potato@1.0.0');
              expect(mockedSpawnSync).not.toBeCalled();
              expect(mockedFork).toHaveBeenNthCalledWith(
                1,
                '/some/silly/path.js',
                [],
                undefined
              );
            });
          });

          describe('when the version requirement is not satisfied', () => {
            beforeEach(() => {
              mockedNpa.mockReturnValue({
                name: 'potato',
                saveSpec: '1.2.3',
              } as unknown as Result);
            });

            it('should fork using the resolved entrypoint after install', () => {
              npf('potato@1.2.3');
              expect(mockedSpawnSync).toHaveBeenNthCalledWith(
                1,
                'npm',
                ['install', '-E', 'potato@1.2.3'],
                {cwd: fakeLazyLoadedDepFolderPath}
              );
              expect(mockedFork).toHaveBeenNthCalledWith(
                1,
                '/some/silly/path.js',
                [],
                undefined
              );
            });
          });
        });

        describe('when the dependency has no entrypoint', () => {
          beforeEach(() => {
            mockedRequireResolve.mockImplementationOnce(doThrowModuleNotFound);
          });

          it('should throw an InternalError', () => {
            try {
              npf('potato@1.2.3');
            } catch (error) {
              expect(error).toBeInstanceOf(InternalError);
              expect((error as InternalError).message).toMatchSnapshot();
            }
          });
        });
      });
    });
  });

  describe('when a packageName cannot be resolved by npa', () => {
    beforeEach(() => {
      mockedNpa.mockReturnValue({
        name: null,
        saveSpec: '1.2.3',
      } as unknown as Result);
    });

    it('should throw an InternalError', () => {
      try {
        npf('potato@1.2.3');
      } catch (error) {
        expect(error).toBeInstanceOf(InternalError);
        expect((error as InternalError).message).toMatchSnapshot();
      }
    });
  });
});
