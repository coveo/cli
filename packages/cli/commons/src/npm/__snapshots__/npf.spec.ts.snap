// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npf when a packageName cannot be resolved by npa should throw an InternalError 1`] = `
"Internal CLI Error, please report to Coveo

Message:
  Package spec potato@1.2.3 seems invalid"
`;

exports[`npf when the packageName can be resolved by npa when the lazyLoad npm project does exists when the dependency is installed when the dependency has no entrypoint should throw an InternalError 1`] = `
"Internal CLI Error, please report to Coveo

Message:
  Package potato does not have an entrypoint."
`;
